### DISCUSSION TITLE
"Global Option" in help text section?

### DISCUSSION THREAD

**Discussion by wayneyaoo** [spectreconsole/spectre.console#643] [Q&A] [answered]
Currently the default help text contains normally 2 sections "USAGE" and "OPTIONS'. Say if I use a base class CommandSetting as global option (e.g., `--verbose`) and all other CommandSettings inherit from it, how can I express this in the help text? is there any "GLOGAL OPTION" section available?

and btw, is there any documentation/API for customizing help text? Couldn't find any in the doc site.

Thanks very much for the great project!

**Comment by nils-a** [comment, marked as answer]
@wayneyaoo there is currently no "GLOBAL OPTIONS" section or anything like this, sorry.

Also, the overall layout of the help text can not be customized.
I.e. the "layout" of `USAGE`, `EXPAMLPES`, `ARGUMENTS` and `OPTIONS` can not be modified.

![image](https://user-images.githubusercontent.com/349188/144046485-8deb2adf-cad7-4f69-958c-b1f2002f70b7.png)
 

* The content of the `EXAMPLES` can be modified using `IConfigurator.AddExample()` or `ICommandConfigurator.WithExample()`, see [*Customizing Command Configurations*](https://spectreconsole.net/cli/commandapp).
* The descriptions for arguments, as well as options can be modified using the `System.ComponentModel.DescriptionAttribute`, see [*Description*](https://spectreconsole.net/cli/settings).

**Reply by wayneyaoo** [reply]
Ha thanks for the reply. 

Is there any plan to add support to customize the help text? Also is there anything I need to note if I wanna contribute? I estimated this to be a medium change... maybe the owner chose not to allow customization for some reason.

**Reply by nils-a** [reply]
Probably everything you need to know before contributing can be found in [CONTRIBUTING.md](https://github.com/spectreconsole/spectre.console/blob/main/CONTRIBUTING.md).

Up to now, there was no plan to make the overall help layout customizable - but we're certainly open to discussions about it.
What would be your idea of making this part customizable?

**Reply by wayneyaoo** [reply]
Thanks for the pointer. Read it.

Actually I have a couple of ideas  more than this to make this project even better (I'm assuming..). To name a few,

1. Allow the lib user to specify whether a branch can be "executed", because currently there's an articulate comment in the code saying that "branch cannot be executed".

2. Allow description for examples. i.e., describe what the example tries to do, which is not always obvious.

3. Interpret verb with no input to be "print help for this verb", which is the topic in this question. [This](https://clig.dev/#help) explains it better. It's of course not definitive guide but such behavior is more preferred sometimes right?

4. Customization of help text. I need to give more thoughts on this but "Global Option" is one candidate here.

Let me know which points here interests you or the owner or any maintainer so that I can open discussion for each one of them. And of course I understand that asking new features with bare hands is mean so I'm happy to implement all of them, in the name of a personal developer. Finally I just feel like there might be some level of architectural change needed to do some of these features properly (well-engineered), or some contradict the existing design philosophy, so feel free to reject if base on your experience some features are too expensive/not preferred to implement.

**Reply by nils-a** [reply]
@wayneyaoo sorry for being so slow to respond currently. I'm always interested in ideas other people have 😄 

I would suggest to open individual discussions on the ideas you have, so each idea can be discussed individually.
